discretenet/problems/water_pipe_enhancement/water_pipe_enhancement.py

```diff
@@ -0,0 +1,252 @@
+            pyo.quicksum(model.yT[t] for t in self.T)
+            - pyo.quicksum(model.x[node1, node2] for node1, node2 in self.graph.edges())
+            == len(self.C)
+        )
+
+        self.model = model
+
+    def get_name(self) -> str:
+        return self.name
+
+    def get_parameters(self):
+        # to be implemented
+        return None
+
+
+class WaterPipeEnhancementGenerator(Generator[WaterPipeEnhancementProblem]):
+    def __init__(
+        self,
+        random_seed: int = 42,
+        path_prefix: Union[str, Path] = None,
+        graph_instance="small_toronto",
+        housing_area_rate=0.01,
+        housing_area_size=3,
+        critical_rate=0.01,
+        water_source_rate=0.005,
+    ):
+        """
+        Initialize the water pipe enhancement problem generator instance
+        following https://doi.org/10.1145/3378393.3402246
+        :param random_seed: The random seed to use
+        :param path_prefix:  Path prefix to pass to instance ``save()`` methods
+            during batch generation. Must be set as a public instance attribute,
+            to be changed by the user after generator instantiation if desired.
+        :param graph_instance: the name of the graph instance to use,
+            the graphs are located under the graphs directory
+        :param housing_area_rate: percentage of nodes that are center of housing areas
+        :param housing_area_size: use k-hop neighbours to define each neighbourhood,
+            k = r_size
+        :param critical_rate: percentage of nodes that are critical customers
+        :param water_source_rate: percentage of nodes that are water sources
+        """
+        super().__init__(random_seed, path_prefix)
+        self.housing_area_rate = housing_area_rate
+        self.housing_area_size = housing_area_size
+        self.critical_rate = critical_rate
+        self.water_source_rate = water_source_rate
+        self.graph_instance = graph_instance
+        self.SR = None
+        self.C = None
+        self.T = None
+        self.name = None
+        self.graph = None
+
+        with open_binary(
+            "discretenet.problems.water_pipe_enhancement.graphs",
+            f"{graph_instance}.gpickle",
+        ) as fd:
+            self.base_graph = pickle.load(fd)
+
+    def generate(self):
+        self.graph = self.base_graph.copy()
+        self.undirected_graph = self.graph.to_undirected()
+
+        self.generateCTR()
+        self.name = "{}_RR{}_rS{}_CR{}_TR{}".format(
+            self.graph_instance,
+            self.housing_area_rate,
+            self.housing_area_size,
+            self.critical_rate,
+            self.water_source_rate,
+        )
+        problem = WaterPipeEnhancementProblem(
+            self.graph,
+            self.undirected_graph,
+            self.SR,
+            self.C,
+            self.T,
+            self.name + "_%d" % self.random_seed,
+        )
+        return problem
+
+    def generateCTR(self):
```

Comment:
As with elsewhere, it would be better to have this take in parameters and return values rather than setting / reading instance attributes. The code will be easier to read and debug that way.

Reply:
Updated